atintegrators/ExactHamiltonianPass.cc

```diff
@@ -0,0 +1,207 @@
+#include "atelem.c"
+#include <string.h>
+#include "atlalib.c"
+
+#define AT_MODE
+#include "track.cc"
+
+struct elem
+{
+    double Length;
+    double BendingAngle;
+    double *PolynomA;
+    double *PolynomB;
+    int MaxOrder;
+    int NumIntSteps;
+    int Type;
+    /* Optional fields */
+    double gK;
+    int MultipoleFringe;
+    double *R1;
+    double *R2;
+    double *T1;
+    double *T2;
+};
+
+static const char * FirstOptionalName = "gK";
+
+#define NUM_FIELDS_2_REMEMBER (sizeof(FieldNames) / sizeof(FieldNames[0]))
+
+void ExactHamiltonianPass(double *r_in, double le,
+                     double *A, double *B,
+                     const double *T1, const double *T2,
+                     const double *R1, const double *R2,
+                     int max_order, int num_int_steps,
+                     double phi, int type,
+                     double gK, int multipole_fringe,
+                     int num_particles)
+{
+    int c, n;
+    double * r6;
+
+    // copy AT structure into pass method structure
+
+    element e = {0};
+
+    for(n = 0; n < max_order; n++)
+    {
+        e.F[2 * n]     = B[n];
+        e.F[2 * n + 1] = A[n];
+    }
+    
+    e.L = le;
+    e.phi = phi;
+    e.gK = gK;
+    e.nF = max_order;
+    e.slices = num_int_steps;
+    e.type = type;
+    e.do_multipole_fringe = multipole_fringe;
+
+    for(c = 0; c<num_particles; c++)
+    {
+        r6 = r_in+c*6;	
+        if(!atIsNaN(r6[0]))
+        {   
+            /* misalignment at entrance */
+            if(T1)
+                ATaddvv(r6,T1);
+            if(R1)
+                ATmultmv(r6,R1);
+            
+            track_element(r6, &e);
+            
+            /* misalignment at exit */
+            if(R2)
+                ATmultmv(r6,R2);
+            if(T2)   
+                ATaddvv(r6,T2);
+        }
+    }
+}
+
+#if defined(MATLAB_MEX_FILE) || defined(PYAT)
+
+ExportMode struct elem *trackFunction(const atElem *ElemData, struct elem *Elem,
+                               double *r_in, int num_particles,
+                               struct parameters *Param)
+
+{
+    double le, bending_angle;
+    double *polynom_a, *polynom_b;
+    long max_order, num_int_steps, type, multipole_fringe;
+    double *R1, *R2, *T1, *T2;
+    double phi, gK;
+    le = atGetDouble(ElemData,"Length"); check_error();
+    bending_angle = atGetDouble(ElemData,"BendingAngle"); check_error();
+    polynom_a = atGetDoubleArray(ElemData,"PolynomA"); check_error();
+    polynom_b = atGetDoubleArray(ElemData,"PolynomB"); check_error();
+    max_order = atGetLong(ElemData, "MaxOrder"); check_error();
+    num_int_steps = atGetLong(ElemData, "NumIntSteps"); check_error();
+    type = atGetLong(ElemData, "Type"); check_error();
+    multipole_fringe = atGetLong(ElemData, "Type"); check_error();
```

Comment:
The `"Type"` attribute appears twice, for both `type` and `multipole_fringe` variables. The 2nd time, it should be something like `"MultipleFringe"`, I suppose…